### PR TITLE
estimated transaction fee

### PR DESCRIPTION
**Pull request by isaacJChen**
Note: This change will make the previous versions (currently just 0.6.1 since we deleted everything else) break. Will require another genesis reset.

@isaacJChen update documentation.

### CODE REVIEW

File: lib/bitcoin/BitcoinClient.ts [hunk 1]
```diff
@@ -8,6 +8,7 @@ import nodeFetch, { FetchError, Response, RequestInit } from 'node-fetch';
 import ReadableStream from '../common/ReadableStream';
 import { Address, crypto, Networks, PrivateKey, Script, Transaction, Unit } from 'bitcore-lib';
 import { IBlockInfo } from './BitcoinProcessor';
+import BitcoinDataTransactionModel from './models/BitcoinDataTransactionModel';
```
Comment (mudiali) [inline lib/bitcoin/BitcoinClient.ts]: Sort please.

File: lib/bitcoin/BitcoinClient.ts [hunk 2]
```diff
@@ -28,7 +29,8 @@ export default class BitcoinClient {
     bitcoinRpcPassword: string | undefined,
     bitcoinWalletImportString: string,
     private requestTimeout: number,
-    private requestMaxRetries: number) {
+    private requestMaxRetries: number,
+    private transactionFeeMarkupPercentage: number) {
```
Comment (mudiali) [inline lib/bitcoin/BitcoinClient.ts]: One concern that I have is that this percentage is only applied to the 'sidetree' transactions. Any other transactions such as lock/relock etc don't honor this percentage. So we need to decide between:
1. Either every transaction has this markup.
1. Or rename this variable to denote that this markup is ONLY for the sidetree transactions.

File: lib/bitcoin/BitcoinProcessor.ts [hunk 3]
```diff
@@ -214,40 +215,38 @@ export default class BitcoinProcessor {
...
+   * @param minFee The minimum fee to be paid for this transaction.
    */
-  public async writeTransaction (anchorString: string, fee: number) {
-    console.info(`Fee: ${fee}. Anchoring string ${anchorString}`);
-    // ----
-    // Issue #347 opened to track the investigation for this hardcoded value.
-    // Ensure that we are always paying this minimum fee.
-    fee = Math.max(fee, 1000);
-    // ----
+  public async writeTransaction (anchorString: string, minFee: number) {
```
Comment (mudiali) [inline lib/bitcoin/BitcoinProcessor.ts]: This changes the contract between core and bitcoin API and needs documentation updates. https://github.com/decentralized-identity/sidetree/blob/master/docs/implementation.md#write-a-sidetree-transaction

File: lib/bitcoin/IBitcoinConfig.ts [hunk 4]
```diff
@@ -16,4 +16,5 @@ export interface IBitcoinConfig {
   requestTimeoutInMilliseconds: number | undefined;
   requestMaxRetries: number | undefined;
   transactionPollPeriodInSeconds: number | undefined;
+  transactionFeeMarkupPercentage: number | undefined;
```
Comment (mudiali) [inline lib/bitcoin/IBitcoinConfig.ts]: 1. We need a corresponding change in the ion repo to include this new parameter.
1. Can we keep this variable as required? And by default set it to 0?

File: lib/bitcoin/BitcoinClient.ts (hunk 2 above)
Comment (isaacJChen) [reply lib/bitcoin/BitcoinClient.ts]: Will rename to only apply to sidetree transactions as discusses yesterday.

File: docs/implementation.md [hunk 5]
```diff
@@ -402,7 +402,7 @@ POST /transactions
 #### Request body schema
 ```json
 {
-  "fee": "A number representing the transaction fee to be paid to write this transaction to the blockchain.",
+  "minFee": "A number representing the minimum transaction fee to be paid to write this transaction to the blockchain. The actual fee paid can fluctuate deterministically.",
```
Comment (mudiali) [inline docs/implementation.md]: How about ````minimumFree````

Comment (mudiali) [inline docs/implementation.md]: Just a reminder that this requires a corresponding change in ion repo as well.

Comment (mudiali) [inline docs/implementation.md]:
```suggestion
  "minFee": "A number representing the minimum transaction fee to be paid to write this transaction to the blockchain. The actual fee is dynamically calculated and can be higher than this amount (but not lower).",
```

Comment (isaacJChen) [reply docs/implementation.md]: Good call. I created an issue so we remember what we've done when we update ion

https://github.com/decentralized-identity/ion/issues/56

Comment (isaacJChen) [reply docs/implementation.md]: Is it free though? JK, I like it, will change.

File: lib/bitcoin/BitcoinClient.ts [hunk 6]
```diff
@@ -1,5 +1,6 @@
 import * as httpStatus from 'http-status';
 import BitcoinBlockModel from './models/BitcoinBlockModel';
+import BitcoinDataTransactionModel from './models/BitcoinDataTransactionModel';
```
Comment (thehenrytsai) [inline lib/bitcoin/BitcoinClient.ts]: Makes more sense to me if it was `BitcoinTransactionDataModel`.

File: lib/bitcoin/IBitcoinConfig.ts [hunk 7]
```diff
@@ -16,4 +16,5 @@ export interface IBitcoinConfig {
   requestTimeoutInMilliseconds: number | undefined;
   requestMaxRetries: number | undefined;
   transactionPollPeriodInSeconds: number | undefined;
+  sidetreeTransactionFeeMarkupPercentage: number;
```
Comment (thehenrytsai) [inline lib/bitcoin/IBitcoinConfig.ts]: Consider renaming to just `transactionFeeMarkupPercentage`.

File: lib/core/models/Config.ts [hunk 8]
```diff
@@ -4,7 +4,7 @@
 export default interface Config {
   batchingIntervalInSeconds: number;
   blockchainServiceUri: string;
-  transactionFeeMarkupPercentage: number;
+  sidetreeTransactionFeeMarkupPercentage: number;
```
Comment (thehenrytsai) [inline lib/core/models/Config.ts]: Sanity: not needed?

File: lib/core/versions/latest/FeeManager.ts [hunk 9]
```diff
@@ -17,7 +17,7 @@ export default class FeeManager {
    *
    * @throws if the number of operations are <= 0.
    */
-  public static computeTransactionFee (normalizedFee: number, numberOfOperations: number, feeMarkupPercentage: number): number {
+  public static computeTransactionFee (normalizedFee: number, numberOfOperations: number): number {
```
Comment (thehenrytsai) [inline lib/core/versions/latest/FeeManager.ts]: Consider:
```suggestion
  public static computeMinimalTransactionFee (normalizedFee: number, numberOfOperations: number): number {

```

File: tests/json/config-test.json [hunk 10]
```diff
@@ -6,6 +6,5 @@
   "batchingIntervalInSeconds": 10,
   "observingIntervalInSeconds": 10,
   "maxConcurrentDownloads": 20,
-  "mongoDbConnectionString": "mongodb://localhost:27017/",
-  "transactionFeeMarkupPercentage": 10
```
Comment (thehenrytsai) [inline tests/json/config-test.json]: Sanity: don't forget to remove this from ION (and internal ION) repo also.

Comment (isaacJChen) [reply tests/json/config-test.json]: Yes, I have an issue for that

File: lib/core/models/Config.ts (hunk 8 above)
Comment (isaacJChen) [reply lib/core/models/Config.ts]: yes

**Pull request merged**